### PR TITLE
Browser support

### PR DESCRIPTION
**Pull request by EgoAleSum**
As per issue #25, this patch makes shortId able to run on a browser.

Notes:
1. On a browser, just include "shortId.js", which exposes the global shortId() method (and shortId.generate(), etc)
2. The file can also be included in another project that supports CommonJS/Browserify in the normal way: `require('shortid')`. In this case, no global object is exported.
3. Node.js crypto module is not used on the browser. Instead, we use the Web Crypto API when available, with a fallback to Math.random()
4. All the tasks are done with grunt (browserify and uglify).
5. browserify has two "transforms": envify because lib/shortid.js uses the NODE_UNIQUE_ID environmental variable (which is set to 0 always, here) and derequire so the produced "shortId.js" file can be loaded by other AMD/CommonJS loaders
6. I ported the test for shortId to run on the browser with mocha.

Please note that I don't have access to a machine with lots of browsers to test on. I've run tests also disabling the Web Crypto APIs, but I can't test on IE 10 or other older browser from here.

Also, I leave it to you to update the docs :)

### CODE REVIEW

**Comment by dylang** [comment]
Thanks for doing this!

Can you move the compiled code to a `dist` directory and add `/dist/` to `.npm_ignore`.  I'd also prefer if we `.gitignore` `/dist/` and require devs to use `browserify` or `webpack` - what do you think (and anyone else who is hoping for browser support)?

Can we share the same tests between browser and Node? I'm not a fan of the duplication and not running all of the tests on the browser.

What do you think about using something like PhantomJS to show the browser test pass? It would require a test runner like [Karma](http://karma-runner.github.io/). This way `travis` can run both the Node and the browser tests.

File: Gruntfile.js [hunk 1]
```diff
@@ -4,7 +4,65 @@ module.exports = function (grunt) {
...
+    var envifyTransform = envify({NODE_UNIQUE_ID: 0});
+
+    // String to be prepended to source files
+    var banner = "/*! shortId.js\n https://github.com/dylang/shortid\n Copyright (c) Dylan Greene\n License: MIT +no-false-attribs License\n*/\n";
+
     grunt.initConfig({
+        browserify: {
+            dist: {
+                src: ['lib/shortid.js'],
+                dest: 'shortId.js'
```
Comment (dylang) [inline Gruntfile.js]: how about writing all files to a `/dist/` dir. Also lets just use lowercase for the filenames so a typo doesn't just happen to work on Macs but fails when pushed to a Linux server.

Comment (EgoAleSum) [inline Gruntfile.js]: I'm ok with that, but in all your samples you always write stuff like `var shortId = require('shortid')`, so I wanted to be consistent (the name of the globally exported variable is shortId).

Btw, Macs are case-sensitive as well.

**Comment by EgoAleSum** [comment]
Ok, so:

1. I believe it's a good idea to have the compiled files available in the git repo. Not everyone likes CommonJS, and many users will simply want to grab the .js file to load with a HTML script tag. At the end, however, it's your call.

2. To share tests... I agree, but then you should separate the actual tests into another file (without the "require()" calls).

3. I'm not familiar with karma, sorry

**Comment by dylang** [comment]
Thanks for doing this work @EgoAleSum, it was extremely helpful in getting browser support finally in `shortid`. 

I made some changes that I thought you might want to be aware of:

* I'm not including the browserfied bundle in the repo. I don't like committing built code because it makes 
diffs messy and it's too easy for me to forget to rebuild every time I accept a PR. 
* I include all Node test in the browser tests by browserify the tests as a separate bundle.
* I simplified the _browser vs node_ logic by using the `browser` field in the `package.json` to replace certain modules with browser-specific ones during build time. This means browsers won't get any code that references `crypto` or `process.env` automatically, which simplified the browserify config. This was also required so that other projects can correctly browserify/webpack this project.

Thanks again for your help and sorry it took so long to get this in. My two young kids, some sickness, getting-old injuries, travel, and work have made it difficult to find time to work on my open source projects. I had planned on starting my taxes today while the kids napped, but I focused on this instead.

**Pull request closed without merging**